deckhand/engine/layering.py

```diff
@@ -0,0 +1,292 @@
+        defined by its `layeringDefinition`.
+
+        :returns: the list of rendered documents (does not include layering
+            policy document).
+        """
+        # ``rendered_data_by_layer`` agglomerates the set of changes across all
+        # actions across each layer for a specific document. Use copy
+        # to prevent global data from being updated referentially.
+        rendered_data_by_layer = {}
+
+        # NOTE(fmontei): ``global_docs`` represents the topmost documents in
+        # the system. It should probably be impossible for more than 1
+        # top-level doc to exist, but handle multiple for now.
+        global_docs = [doc for doc in self.layered_docs
+                       if doc.get_layer() == self.layer_order[0]]
+
+        for doc in global_docs:
+            layer_idx = self.layer_order.index(doc.get_layer())
+            rendered_data_by_layer[layer_idx] = doc.all_data
+
+            # Keep iterating as long as a child exists.
+            for child in doc.get_children(nested=True):
+
+                # Retrieve the most up-to-date rendered_data (by
+                # referencing the child's parent's data).
+                child_layer_idx = self.layer_order.index(child.get_layer())
+                rendered_data = rendered_data_by_layer[child_layer_idx - 1]
+
+                # Apply each action to the current document.
+                actions = child.get_actions()
+                for action in actions:
+                    rendered_data = self._apply_action(
+                        action, child.all_data, rendered_data)
+
+                # Update the actual document data if concrete.
+                if not child.is_abstract():
+                    self.layered_docs[self.layered_docs.index(child)][
+                        'data'] = rendered_data['data']
+
+                # Update ``rendered_data_by_layer`` for this layer so that
+                # children in deeper layers can reference the most up-to-date
+                # changes.
+                rendered_data_by_layer[child_layer_idx] = rendered_data
+
+            if 'children' in doc:
+                del doc['children']
+
+        return [d.all_data for d in self.layered_docs]
+
+    def _apply_action(self, action, child_data, overall_data):
+        """Apply actions to each layer that is rendered.
+
+        Supported actions are:
+
+            * `merge` - a "deep" merge that layers new and modified data onto
+              existing data
+            * `replace` - overwrite data at the specified path and replace it
+              with the data given in this document
+            * `delete` - remove the data at the specified path
+
+        Requirements:
+
+            * The path must be present in both ``child_data`` and
+              ``overall_data`` (in both the parent and child documents).
+        """
+        method = action['method']
+        if method not in self.SUPPORTED_METHODS:
```

Comment:
Add TODO saying this should defer to schema validation. Validation should've been done by now,